### DISCUSSION TITLE
Annoying little issues

### DISCUSSION THREAD

**Discussion by hydroper** [airsdk/Adobe-Runtime-Support#2522] [General]
I'm having some annoying issues which results in bugs:

## ActionScript

- `new Vector.<String>(['xxx']).length === 0`
  - But `new <String>['xxx'].length === 1`
- `iterableArrayOrVector.map(function(a) ...)`
  - You're forced to specify rest arguments, since `map` receives more than one argument; like `function(a, ..._)`.

## ASDoc

- ASDoc unclosed tag errors don't specify the source location, so you've to hunt the source file yourself.
- I'm getting a bytecode error for a specific library:
```text
C:\Users\matheus-pc\Documents\adobe-air-projects\rialight\api\com.rialight.intl>aasdoc --library-path+=libs --doc-source
s+=src -o=build/docs
Using Flex SDK: C:\Dev\AIRSDK_Windows\AIRSDK_50.0.1.2

C:\Users\matheus-pc\Documents\adobe-air-projects\rialight\api\com.rialight.intl>java -Dsun.io.useCanonCaches=false -Xms32m -Xmx512m -Dflexcompiler="C:\Dev\AIRSDK_Windows\AIRSDK_50.0.1.2\bin\.." -Dflexlib="C:\Dev\AIRSDK_Windows\AIRSDK_50.0.1.2/frameworks" -Dflex.compiler.theme= -Xbootclasspath/p:"C:\Dev\AIRSDK_Windows\AIRSDK_50.0.1.2/asdoc/xalan.jar" -classpath "C:\Dev\AIRSDK_Windows\AIRSDK_50.0.1.2/lib/legacy/asdoc.jar" flex2.tools.ASDoc -compiler.fonts.local-fonts-snapshot= +configname=air --library-path+=libs --doc-sources+=src -o=build/docs
Loading configuration file C:\Dev\AIRSDK_Windows\AIRSDK_50.0.1.2\frameworks\air-config.xml
C:\Users\matheus-pc\Documents\adobe-air-projects\rialight\api\com.rialight.intl\libs\com.rialight.util.swc(com.rialight.util:Map): Error: abc bytecode decoding failed.

C:\Users\matheus-pc\Documents\adobe-air-projects\rialight\api\com.rialight.intl\libs\com.rialight.util.swc(com.rialight.util:Set): Error: abc bytecode decoding failed.
```

## Etc.

There are probably things I'm missing here. If I remember any, I'll update the post.

**Comment by ajwfrost** [comment]
Hi

`new Vector.<String>(['xxx'])` looks invalid to me, perhaps it should throw an error? Documentation says that the [Vector constructor](https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/Vector.html#Vector()) takes an optional length value; it looks like you are actually thinking of the [Vector function](https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/package.html#Vector()) - so then it should work if you do just `var x = Vector.<String>(['xxx'])` i.e. remove the `new` keyword.

For the `map` function, the Function callback takes three arguments according to the documentation, I guess you would need to ensure the function prototype fits with this but I'm not sure it's necessary to use `...`?

ASDoc isn't something we've looked at (yet!) - I'll check whether we have the source code for it...


thanks

**Reply by hydroper** [reply]
Thanks, Andrew! Yeah, seems like `Vector` takes a number... Calling it with an `Array`, but without `new`, works. I'd suggest for `Vector` to throw an `ArgumentError` then, since that leads to silent bugs.

**Reply by ajwfrost** [reply]
> I'd suggest for `Vector` to throw an `ArgumentError` then, since that leads to silent bugs.

Agreed! Complete lack of type checking: https://github.com/adobe/avmplus/blob/master/core/VectorClass-impl.h#L90 ...

Simple to change it:
![image](https://user-images.githubusercontent.com/6400804/224685755-9326e30c-c0b3-4001-948b-415a1206392f.png)


thanks
